### ISSUE TITLE
Potential bug of logn_scale

### ISSUE THREAD

**Issue by SkyAndCloud** [DAMO-NLP-SG/CLEX#1]
Hi, I just found that there is a potential bug for the logn implementation in this repo. As shown in [line](https://github.com/DAMO-NLP-SG/CLEX/blob/a48602f9cc8fbc6ffad58a1d55d19b027e793f1e/CLEX/modeling_llama.py#L304C38-L304C38), the scale factor is `math.log(k_len) / math.log(train_len)` for each q. However, in Su's [blog](https://kexue.fm/archives/8823), it should be `torch.arange(k_len).log() / math.log(train_len)`. Its implementation can also be found at [ReRoPE](https://github.com/bojone/rerope/blob/156ec82065e0d0e65b8deac56ff48427def7fc98/ntk_patch.py#L24C14-L24C14)

**Comment by guanzhchen** [comment]
Thank you for pointing out this issue!  The log_n scaling in our experiments is an inference-time trick. We don't use any log_n scaling during training, so feel free to stop it during inference if you have some concerns. 

While the log_n scaling in our implementation is indeed different from that in the ReRoPE, as we solely aim to apply constant scaling.  We don't encourage using the original implementation, as it seems to further amplify the local (tail) attention scores after softmax normalization. This may be helpful for reducing the next-token prediction loss (and thus lower ppl on language modelling), however, we deem it would lose partial long-distance information like alibi or sliding window.

We just conduct an experiment on MultiFieldQA-en task in Longbench, using the implementation in ReRoPE (denoted as ReRoPE log_n). The results are listed in the table:
|              | MultiFieldQA-en |
| ------------ | -------- |
| ReRoPE log_n | 42.69    |
| no log_n     | 43.14    |
| Our log n    | 44.62    |

We may use a "wrong" log_n scaling implementation although, we found our implementation is helpful while ReRoPE implementation slightly harms the performance. This performance drop may originate from the ununiform log_n scaling as we mentioned above.

We guess there should be a constant (uniform) scaling operation, as the attention may assign some scores to some unnecessary tokens, which would over-distribute the attention scores in long-context setting. While uniformly upscaling the attention scores would benefit the focus on more related tokens (with higher scores) after softmax normalization.  Our naive log_n implementation is probably not the best (e.g., [Yarn](https://arxiv.org/abs/2309.00071) proposes another scaling trick),  so we would consider taking some time to explore more about it.

Feel free to contact us if have any more questions.
